From 31c0b06a652121bb58683010a89687b1c5058da3 Mon Sep 17 00:00:00 2001
From: Gonzalo Odiard <godiard@gmail.com>
Date: Fri, 1 Nov 2013 12:52:19 -0300
Subject: [PATCH 6/8] Stop all the activities if tried to shutdown after 30
 seconds

When a activity crashed or can't close properly, like the actual version
of Terminal Sugar can't logout or shutdown the computer.
This patch wait 30 seconds and request all the activities to close.

Signed-off-by: Gonzalo Odiard <gonzalo@laptop.org>
---
 src/jarabe/model/session.py | 6 ++++++
 src/jarabe/model/shell.py   | 5 +++++
 2 files changed, 11 insertions(+)

diff --git a/src/jarabe/model/session.py b/src/jarabe/model/session.py
index bb42dd6..0620252 100644
--- a/src/jarabe/model/session.py
+++ b/src/jarabe/model/session.py
@@ -62,11 +62,17 @@ class SessionManager(GObject.GObject):
     def initiate_shutdown(self, logout_mode):
         self._logout_mode = logout_mode
         self.shutdown_signal.emit()
+
+        # if sugar can't close in 30 seconds, try this
+        GObject.timeout_add_seconds(30, self.__close_activities)
         self.session.initiate_shutdown()
 
     def __shutdown_completed_cb(self, session):
         GObject.timeout_add_seconds(self.SHUTDOWN_TIMEOUT, self._try_shutdown)
 
+    def __close_activities(self):
+        self._shell_model.stop_all_activities()
+
     def _try_shutdown(self):
         if len(self._shell_model) > 0:
             self._shutdown_tries += 1
diff --git a/src/jarabe/model/shell.py b/src/jarabe/model/shell.py
index f67e31d..4cf3ef5 100644
--- a/src/jarabe/model/shell.py
+++ b/src/jarabe/model/shell.py
@@ -753,6 +753,11 @@ class ShellModel(GObject.GObject):
     def has_modal(self):
         return self._modal_dialogs_counter > 0
 
+    def stop_all_activities(self):
+        # This method should be used only at sugar shutdown
+        for home_activity in self._activities:
+            home_activity.stop()
+
 
 def get_model():
     global _model
-- 
1.8.1.4

